gymnasium_robotics/envs/adroit_hand/adroit_pen.py

```diff
@@ -277,31 +274,39 @@ def step(self, a):
             - self.data.site_xpos[self.tar_b_site_id]
         ) / self.tar_length
 
-        # pos cost
-        dist = np.linalg.norm(obj_pos - desired_loc)
-        reward = -dist * (not self.sparse_reward)
-        # orien cost
+        # compute the sparse reward variant first
+        goal_distance = np.linalg.norm(obj_pos - desired_loc)
         orien_similarity = np.dot(obj_orien, desired_orien)
-        reward += orien_similarity * (not self.sparse_reward)
+        goal_achieved = (
+            True if (goal_distance < 0.075 and orien_similarity > 0.95) else False
+        )
+        reward = 100.0 if goal_achieved else -0.1
+        goal_failed = obj_pos[2] < 0.075
 
-        # bonus for being close to desired orientation
-        if dist < 0.075 and orien_similarity > 0.9:
-            reward += 10
-        if dist < 0.075 and orien_similarity > 0.95:
-            reward += 50
+        # override reward if not sparse reward
+        if not self.sparse_reward:
+            reward = -goal_distance + orien_similarity
 
-        # penalty for dropping the pen
-        terminated = False
-        if obj_pos[2] < 0.075:
-            reward -= 5
-            terminated = True
+            # bonus for being close to desired orientation
+            if goal_distance < 0.075 and orien_similarity > 0.9:
+                reward += 10
+            if goal_distance < 0.075 and orien_similarity > 0.95:
+                reward += 50
 
-        goal_achieved = True if (dist < 0.075 and orien_similarity > 0.95) else False
+            # penalty for dropping the pen
+            if obj_pos[2] < 0.075:
+                reward -= 5
 
         if self.render_mode == "human":
             self.render()
 
-        return obs, reward, terminated, False, dict(success=goal_achieved)
+        return (
+            obs,
+            reward,
+            goal_failed or goal_achieved,
```

Comment:
terminated `False`

Reply:
Reverted for the rest, but would you want the same to happen for Pen? If the pen drops (goal_failed), it becomes impossible for the agent to recover.

Reply:
That is a great idea. Can we add this as an issue enhancement and implement it in a future release with an environment bump? I'd like to generate the Minari datasets with the original D4RL logic